internal/app/ntpd/pkg/ntp/ntp.go

```diff
@@ -5,93 +5,110 @@
 package ntp
 
 import (
-	"errors"
+	"fmt"
 	"log"
 	"math/rand"
 	"syscall"
 	"time"
 
 	"github.com/beevik/ntp"
-)
-
-// https://access.redhat.com/solutions/39194
-// Using the above as reference for setting min/max
-const (
-	MaxPoll = 1000
-	MinPoll = 20
+	"github.com/hashicorp/go-multierror"
 )
 
 // NTP contains a server address
-// and the most recent response from a query
 type NTP struct {
-	Server   string
-	Response *ntp.Response
+	Server  string
+	MinPoll time.Duration
+	MaxPoll time.Duration
+	Retry   int
 }
 
 // NewNTPClient instantiates a new ntp client for the
 // specified server
-func NewNTPClient(server string) *NTP {
-	return &NTP{Server: server}
+func NewNTPClient(opts ...Option) (*NTP, error) {
+	ntp := defaultOptions()
+
+	var result *multierror.Error
+	for _, setter := range opts {
+		result = multierror.Append(setter(ntp))
+	}
+
+	return ntp, result.ErrorOrNil()
 }
 
 // Daemon runs the control loop for query and set time
 // We dont ever want the daemon to stop, so we only log
 // errors
 func (n *NTP) Daemon() (err error) {
-	rando := rand.New(rand.NewSource(time.Now().UnixNano()))
-	ticker := time.NewTicker(time.Duration(rando.Intn(MaxPoll)+MinPoll) * time.Second)
 
 	// Do an initial hard set of time to ensure clock skew isnt too far off
 	var resp *ntp.Response
-	resp, err = n.Query()
-	if err != nil {
+	if resp, err = n.Query(); err != nil {
 		log.Printf("error querying %s for time, %s", n.Server, err)
 		return err
 	}
-	n.Response = resp
 
-	if err = n.SetTime(resp); err != nil {
+	if err = setTime(resp.Time); err != nil {
```

Comment:
should we always use `ClockOffset` to set the time? (not sure here, but I don't see a difference between first and subsequent attempts, as they all just call settimeofday). 

If we used another advanced API to slowly adjust for clock offset, it would make more sense to have different implementation for the first iteration and subsequent iterations. But as in the end it boils down to `setTime` in both cases, the different is only how we calculate timestamp?

Reply:
there's [syscall.Adjtimex](https://golang.org/pkg/syscall/#Adjtimex) // http://man7.org/linux/man-pages/man2/adjtimex.2.html which i'd probably need to read up more on.